src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,361 @@
+                                                                                      _amountsAndSizes,
+                                                                                      _targetAmount,
+                                                                                      _targetAmountPlusOffset,
+                                                                                      _availableTotalSize)
+{
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    iterations = 0;
+    #endif
+}
+
+CCoinsSelectionSlidingWindow::~CCoinsSelectionSlidingWindow() {
+}
+
+void CCoinsSelectionSlidingWindow::Reset()
+{
+    CCoinsSelectionAlgorithm::Reset();
+
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    iterations = 0;
+    #endif
+}
+
+void CCoinsSelectionSlidingWindow::Solve()
+{
+    if (!isRunning)
+    {
+        isRunning = true;
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        uint64_t microsecondsBefore = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+        #endif
+        size_t tempTotalSize = 0;
+        CAmount tempTotalAmount = 0;
+        unsigned int tempTotalSelection = 0;
+        int exclusionIndex = maxIndex;
+        int inclusionIndex = maxIndex;
+        for (; inclusionIndex >= 0; --inclusionIndex)   
+        {
+            if (stopRequested)
+            {
+                break;
+            }
+            tempSelection[inclusionIndex] = true;
+            tempTotalSize += sizes[inclusionIndex];
+            tempTotalAmount += amounts[inclusionIndex];
+            tempTotalSelection += 1;
+            while (tempTotalSize > availableTotalSize ||
+                   tempTotalAmount > targetAmountPlusOffset)
+            {
+                tempSelection[exclusionIndex] = false;
+                tempTotalSize -= sizes[exclusionIndex];
+                tempTotalAmount -= amounts[exclusionIndex];
+                tempTotalSelection -= 1;
+                --exclusionIndex;
+            }
+            if (tempTotalAmount >= targetAmount)
+            {
+                optimalTotalSize = tempTotalSize;
+                optimalTotalAmount = tempTotalAmount;
+                optimalTotalSelection = tempTotalSelection;
+                for (int index = 0; index < problemDimension; ++index)
+                {
+                    optimalSelection[index] = tempSelection[index];
+                }
+                break;
+            }
+        }
+        #if COINS_SELECTION_ALGORITHM_PROFILING
+        iterations = (maxIndex + 1 - inclusionIndex) + (maxIndex - exclusionIndex);
+        executionMicroseconds = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch()).count() - microsecondsBefore;
+        #endif
+        if (stopRequested == false)
+            hasCompleted = true;
```

Comment:
This is wrong: it changes hasCompleted to true only if not stopRequested.
So this may happen : run one cycle without requesting stop (which will set to hasCompleted to true) then invoke another cycle and this time request stop : as a result hasCompleted is still true.
As a safety measure you should :
- make hasCompleted atomic (is written by one thread and read in another)
- reset hasCompleted to false at the same point where you set isRunning = true
- change this to `hasCompleted = !stopRequested`

Reply:
I agree with your suggestion.
Modified accordingly (with additional review of reset method).